components/server/src/server.rs

```diff
@@ -1259,23 +1259,41 @@ impl TiKVServer<RocksEngine> {
         flow_listener: engine_rocks::FlowListener,
     ) -> (Engines<RocksEngine, RocksEngine>, Arc<EnginesResourceInfo>) {
         let block_cache = self.config.storage.block_cache.build_shared_cache();
+        let shared_block_cache = block_cache.is_some();
         let env = self
             .config
             .build_shared_rocks_env(self.encryption_key_manager.clone(), get_io_rate_limiter())
             .unwrap();
 
-        // Create raft engine.
-        let raft_db_path = Path::new(&self.config.raft_store.raftdb_path);
+        let mut raft_data_state_machine = RaftDataStateMachine::new(
+            &self.config.raft_engine.config().dir,
+            &self.config.raft_store.raftdb_path,
+        );
+        let dump_source = raft_data_state_machine.before_open_target();
```

Comment:
Seems it's okay to do the check after opening

Reply:
Opening first will create a non-empty directory. `target` and `source` coexist.